test/framework/src/main/java/org/elasticsearch/test/ClusterServiceUtils.java

```diff
@@ -179,4 +187,30 @@ public static void setAllElapsedMillis(ClusterStatePublicationEvent clusterState
         clusterStatePublicationEvent.setMasterApplyElapsedMillis(0L);
     }
 
+    public static void awaitClusterState(Logger logger,
+                                         Predicate<ClusterState> statePredicate,
+                                         ClusterService clusterService) throws Exception {
+        final ClusterStateObserver observer =
+            new ClusterStateObserver(clusterService, logger, clusterService.getClusterApplierService().threadPool().getThreadContext());
+        if (statePredicate.test(observer.setAndGetObservedState()) == false) {
+            final PlainActionFuture<Void> future = PlainActionFuture.newFuture();
+            observer.waitForNextChange(new ClusterStateObserver.Listener() {
+                @Override
+                public void onNewClusterState(ClusterState state) {
+                    future.onResponse(null);
+                }
+
+                @Override
+                public void onClusterServiceClose() {
+                    future.onFailure(new NodeClosedException(clusterService.localNode()));
+                }
+
+                @Override
+                public void onTimeout(TimeValue timeout) {
+                    future.onFailure(new TimeoutException());
```

Comment:
```suggestion
                    assert false : "onTimeout called with no timeout set";
```

Reply:
This isn't quite correct I think, the default timeout on the observer is 60s isn't it?

Reply:
Oh right so it is. I saw lots of `@Nullable` things, missed that one place. Still, we don't want the observer to timeout do we? We should pass `null` to avoid enqueueing the timeout task.

Reply:
I guess it doesn't matter much since we only wait on the future for 30s so we won't ever see those 60s elapse. But you're right, much cleaner to not enqueue any timeout :)